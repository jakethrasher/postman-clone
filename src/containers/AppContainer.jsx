import React, { Component } from 'react';
import ResultsDisplay from '../display/ResultsDisplay';
import { deleteData, getData, postData, updateData } from '../services/apiUtils';
import Form from '../components/form/Form';
import HistoryDisplay from '../components/history/HistoryDisplay';
import styles from './container.css';

export default class AppContainer extends Component {
    state={
        loading:false,
        method:'GET',
        url:'',
        response:'',
        reqBody:'',
        history:[],
    }

    handleMethodChange =({target})=>{
        this.setState({
            method:target.value,
        })
    }

    handleInputChange = (e) =>{
        const {name,value} = e.target;
       
        this.setState({
           [name]:value
        })
    }

    handleSubmit= async (e) =>{
        e.preventDefault()
        this.setState({loading:true})
        const {method,url,reqBody,history} = this.state;
    
        if(method === 'GET'){
            const response = await getData(url);
            this.setState((prevState)=>({
                history:[...prevState.history,{method,url}],
                response,
                loading:false
            }));
        }else if(method === 'POST'){
            const response = await postData(url,reqBody);
            this.setState((prevState)=>({
                history:[...prevState.history,{method,url}],
                response,
                loading:false
            }));
        }else if(method === 'PUT'){
            const response = await updateData(url,reqBody);
            this.setState((prevState)=>({
                history:[...prevState.history,{method,url}],
                response,
                loading:false
            }));
        }else if(method === 'DELETE'){
            const response = await deleteData(url);
            this.setState((prevState)=>({
                history:[...prevState.history,{method,url}],
                response,
                loading:false
            }))
        }
    }
            
    render() {
        
        return (
        <div className={styles.mainContainer}>
            <HistoryDisplay history={this.state.history}/>
            <div className={styles.rightContainer}>
                <Form 
                    handleSubmit={this.handleSubmit} 
                    method={this.state.method} 
                    onMethodChange={this.handleMethodChange}
                    handleInputChange={this.handleInputChange}
                    url={this.state.url}
                    reqBody={this.state.reqBody}/>
                <ResultsDisplay 
                    response={this.state.response} loading={this.state.loading}/>
            </div>
        </div>   
        )
    }
}
        