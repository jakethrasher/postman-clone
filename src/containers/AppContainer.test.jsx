import React from 'react';
import { screen, render, waitFor, getByText, getByLabelText } from '@testing-library/react'
import userEvent  from '@testing-library/user-event'
import AppContainer from './AppContainer';
import {setupServer} from 'msw/node'
import { rest } from 'msw';
import { exact } from 'prop-types';

const server = setupServer(
    rest.get('https://thesimpsonsquoteapi.glitch.me/quotes', (req, res, ctx) => {
        return res(ctx.json([
            {
              "quote": "These are my only friends...grown-up nerds like Gore Vidal. And even he's kissed more boys than I ever will.",
              "character": "Lisa Simpson",
              "image": "https://cdn.glitch.com/3c3ffadc-3406-4440-bb95-d40ec8fcde72%2FLisaSimpson.png?1497567512083",
              "characterDirection": "Right"
            }
          ]))
    })
)
describe('AppContainer', ()=>{
    beforeAll(()=> server.listen());
    afterAll(()=> server.close());

    it('should render a list of requests made', async () =>{
        render(<AppContainer/>)

        // const ul = await screen.findByRole('list', {name: 'requests'})
        
        const urlBar = screen.getByPlaceholderText('Enter URL')
        userEvent.type(urlBar,'https://thesimpsonsquoteapi.glitch.me/quotes')
        
        const getButton = screen.getByTestId('get-button')
        userEvent.click(getButton);

        const go = await screen.findByRole('button',{name:'go-button'})
        userEvent.click(go)

        
        return waitFor(()=>{
            const ul = screen.getByRole('list', {name:'requests'});
            expect(ul).not.toBeEmptyDOMElement()
            expect(ul).toMatchSnapshot()
        })
    })
})
