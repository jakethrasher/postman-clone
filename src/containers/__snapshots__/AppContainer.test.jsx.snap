// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer should render a list of requests made 1`] = `
<ul
  aria-label="requests"
>
  <li>
    <span>
      GET
       
    </span>
    <span>
      https://thesimpsonsquoteapi.glitch.me/quotes
    </span>
  </li>
</ul>
`;
